src/machine/aot/mod.rs

```diff
@@ -0,0 +1,682 @@
+
+    // A memory segment which contains code loaded from ELF
+    memory: FlatMemory<u64>,
+    labels: HashSet<u64>,
+    sections: Vec<(u64, u64)>,
+    dummy_sections: HashMap<u64, u64>,
+}
+
+impl LabelGatheringMachine {
+    pub fn load(program: &Bytes) -> Result<Self, Error> {
+        let elf = Elf::parse(&program).unwrap();
+        if elf.section_headers.len() > MAXIMUM_SECTIONS {
+            return Err(Error::LimitReached);
+        }
+        let sections: Vec<(u64, u64)> = elf
+            .section_headers
+            .iter()
+            .filter_map(|section_header| {
+                if section_header.sh_flags & u64::from(SHF_EXECINSTR) != 0 {
+                    Some((
+                        section_header.sh_addr,
+                        section_header.sh_addr + section_header.sh_size,
+                    ))
+                } else {
+                    None
+                }
+            })
+            .rev()
+            .collect();
+        // Test no section overlaps with one another.
+        // TODO: find a faster algorithm if this proves to be a bottleneck
+        for (i, (start_a, end_a)) in sections.iter().enumerate() {
+            for (j, (start_b, end_b)) in sections.iter().enumerate() {
+                if i == j {
+                    continue;
+                }
+                if ((start_a >= start_b) && (start_a < end_b))
+                    || ((start_b >= start_a) && (start_b < end_a))
+                {
+                    return Err(Error::OutOfBound);
+                }
+            }
+        }
+        // DefaultCoreMachine is only used here for loading ELF binaries
+        // into memory.
+        let mut inner = DefaultCoreMachine::default();
+        inner.load_elf(&program, false)?;
+
+        Ok(Self {
+            registers: init_registers(),
+            pc: Value::from_u64(0),
+            labels: HashSet::default(),
+            labels_to_test: Vec::new(),
+            memory: inner.take_memory(),
+            sections,
+            dummy_sections: HashMap::default(),
+        })
+    }
+
+    fn read_pc(&self) -> Result<u64, Error> {
+        match &self.pc {
+            Value::Imm(pc) => Ok(*pc),
+            _ => Err(Error::Unexpected),
+        }
+    }
+
+    pub fn gather(&mut self) -> Result<(), Error> {
+        let decoder = build_imac_decoder::<u64>();
+        for i in 0..self.sections.len() {
+            let (section_start, section_end) = self.sections[i];
+            self.labels.insert(section_start);
+            self.pc = Value::from_u64(section_start);
+            let mut start_of_basic_block = true;
+            loop {
+                let pc = self.read_pc()?;
+                if pc >= section_end {
+                    break;
+                }
+                match decoder.decode(&mut self.memory, pc) {
+                    Ok(instruction) => {
+                        start_of_basic_block = is_basic_block_end_instruction(instruction);
```

Comment:
The whole loop looks weird. I think the reason is that in the current iteration, we really are processing the instruction in the next iteration.

Reply:
Pseudo code

```python
pc = section_start
start_of_basic_block = true

while pc < section_end:
    if start_of_basic_block:
        labels.append(pc)

    instruction = decode_at(pc) 
    execute(instruction)
    scan_labels_to_test()
    start_of_basic_block = is_end_of_basic_block(instruction)
    pc = pc + instruction_length(instruction)
```